### PR TITLE
Add context binding test (rough)

### PR DESCRIPTION
**Pull request by wridgeu**
Hi @texttechne,

I have tried to come up with tests based on pre-existing ones but I'm having a hard time coming into the right mindset for testing. I'm not that good in writing tests, even more so when it comes to such "internals" like the context. 

Maybe we can work on that one together(?) and simply have this PR open until it's ready. =)

Additionally I [adjusted the bindContext return type](https://github.com/cpro-js/ui5-mobx/compare/main...wridgeu:feat/add-test?expand=1#diff-a9ea1396d72bad2bce45c77421d9aece5cd05bac1596b76fae46620039670dd8R156) to match what is returned and removed an unnecessary [bindProperty](https://github.com/cpro-js/ui5-mobx/compare/main...wridgeu:feat/add-test?expand=1#diff-e9bb8753f781096a07541d83b2e5474099b9e41d6b354d6fe78cd1977f403a8aL47) call, as it is done within "beforeEach" of the test.

BR, Marco

### CODE REVIEW

**Comment by texttechne** [comment]
Hi @wridgeu,

fixed some stuff first:
* I got compile errors for MobxContextBinding
  * added compiling step to build script as safety net
* export for MobxContextBinding
  * probably not needed anyways, but more consistent with the rest

**Comment by texttechne** [comment]
Hi @wridgeu,

yeah it's definitely not an easy task to test this lib.

And I'm not too sure if the taken approach is the right one or sufficient. I've tried to setup the testing along the path that the initial article / version used. And I only got it done with the help of @petermuessig. Since then I put the topic on hold, so to say. 

I'm actually quite open on how to approach this topic, but here are my 2 cents.

### Unit Tests
I like to have this kind of unit testing as it is setup right now (however, I'm not a Karma fan...): Testing in isolation and as fast as possible. 

Nevertheless, there's a downside to it: These unit tests rely on the implicit knowledge how the model is used by UI5. There's a lot of exposed state, events are triggered, etc, which is not such an easy input-output-relation as one would like to have for unit tests.

In the end, these tests do not suffice to give me enough confidence to say that the lib is working in combination with UI5.

### Integration Tests
Hence, I think we should also add integration tests.

I really liked the wdi5 stuff presented at the UI5con....

File: test/cpro/js/ui5/mobx/MobxModelContextBinding.test.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  let model: MobxModel<TestState>;
+  let binding: MobxContextBinding;
+  let ctx: Context;
+
+  const NEW_VALUE = "one to bind them all";
+
+  beforeEach(() => {
+    state = observable(createTestData());
+    model = new MobxModel(state);
+    ctx = new Context(model, "/complex");
```
Comment (texttechne) [inline test/cpro/js/ui5/mobx/MobxModelContextBinding.test.ts:19]: I've trouble to imagine how this works in a UI5 component. I have a parent component which performs the context binding (`binding="{/complex}"`). And then comes the child complonent performing its binding (`binding="{a}").

Are you sure that `model.bindContext` is the only interaction with the model?

**Comment by wridgeu** [comment]
I understand and am on your side especially in regards to requiring more in-depth knowledge of UI5s internals as well lacking confidence in the current tests. I also thought of "bringing in" actual UI5 controls for better testing but that'd go into the integration testing direction.

> I like to have this kind of unit testing as it is setup right now (however, I'm not a Karma fan...): Testing in isolation and as fast as possible.

True! In regards to Karma (which is now deprecated anyway) we could have a look at https://github.com/ArnaudBuchholz/ui5-test-runner which already found use in Marians https://github.com/marianfoo/ui5-cc-spreadsheetimporter.

wdi5 seems great and quite easy to get started to, though I haven't had the chance to properly make use of it yet (was a lot in SAP BAS previously which just recently got gecko driver support for it). Would you want to have a UI5 application in the same repo, consuming the lib for testing or how'd you approach it?

File: test/cpro/js/ui5/mobx/MobxModelContextBinding.test.ts (hunk 1 above)
Comment (wridgeu) [reply test/cpro/js/ui5/mobx/MobxModelContextBinding.test.ts:19]: The way I thought of it was, that it is more a combination of both context and property/aggr binding when working with UI5 controls as a context will be bound to the parent control, the child-controls have relative bindings which will be resolved against the bound context.

I didn't think about both, parent and child component/control using only the `binding` property. I went through a few context binding scenarios and mostly observed the `[bindContext](https://github.com/SAP/openui5/blob/edb94530aa9d45ec61f79867e04eb3e7d8b529b4/src/sap.ui.core/src/sap/ui/model/ManagedObjectBindingSupport.js#L74)` being called here.

But as already written by you, that's certainly no 100% guarantee.

> And then comes the child component performing its binding (`binding="{a}").

Would that imply the child component (in this specific scenario against the test data) would have some property/aggr binding somewhere with `""` or `"/"` which resolved to the value of 'a'?

**Comment by texttechne** [comment]
First off, I gave you write access to the repo. Welcome!

> True! In regards to Karma (which is now deprecated anyway) we could have a look at https://github.com/ArnaudBuchholz/ui5-test-runner which already found use in Marians https://github.com/marianfoo/ui5-cc-spreadsheetimporter.

ok, created a ticket for that: #5

and also for setting up the sample app: #6
* let's just use a subfolder like "example" (Marian's using it, odata2ts is using it, ...)

and setting up wdi5: #7

Tackle whatever issue you feel like doing!

**Comment by wridgeu** [comment]
> First off, I gave you write access to the repo. Welcome!

Oh wow, appreciate it! Hope I don't break things too much. x) I will try to help where I can and when I have time!

#5 should probably be tackled first.

Should we continue within this branch or would you want to pick some improvements of yours from this branch into main and continue on another one?

**Comment by texttechne** [comment]
Reworked the tests, please check.

According to my humble tests, UI5 first calls `bindContext` and then something like `bindProperty`.
I've changed the test setup accordingly in the `beforeEach` method.

* added (brief) check that the `createBindingContext` function of MobxModel works
* expanded on the failure cases by using `@ts-expect-error` for non valid input
* checked if property and list binding work in combination with the context binding
* expanded on some other binding tests in the same manner

**Comment by texttechne** [comment]
I've also changed the implementation a bit by making the context param of `bindContext` required. The function throws now if no context param is provided.

Played around a bit without passing a proper context and it fails in the aftermath. Hence, the additional constraint which might break things...

**Comment by wridgeu** [comment]
This looks a _lot_ better, awesome!

>I've also changed the implementation a bit by making the context param of `bindContext` required. The function throws now if no context param is provided.

Sounds good. I've seen `null` or an empty Context being passed in for the ODataModel calls in some of UI5s test files but without properly understanding their specific context (scimmed through). As client side only model I have no argument at the moment that would actively speak against this.

> expanded on the failure cases by using @ts-expect-error for non valid input

Didn't know about this, makes much more sense, cool.

> checked if property and list binding work in combination with the context binding

Definitely an important missing piece. Sole context binding, without property or list binding used, is rather a "rarity" I'd say.

**Comment by texttechne** [comment]
> Sounds good. I've seen null or an empty Context being passed in for the ODataModel calls in some of UI5s test files but without properly understanding their specific context (scimmed through). As client side only model I have no argument at the moment that would actively speak against this.

As you said: It's allowed. I've just fixed it and re-adapted the tests.

**Pull request merged**